Nhse 034 orkv.i57 fixcapability

martinsumner opened the pull request:
Capabilities are used to determine if a feature is safe to use, where this is determined by all nodes in the cluster being able to use that capability.

Capability calls are spread around the code.  Some capabilities continue to be check, even though there is no support for a version sufficiently old to not support the capability.

There is a minimal cost to capability checks, but that cost is non-zero.  There are also potential failure scenarios where there may be a fallback to the "wrong" capability.

Unify capability definitions, and where they are frequently requested and backwards-compatibility is no longer possible, hardcode them.

martinsumner committed:
Centralise capability requests

... to make it easier to default in the future

martinsumner committed:
Fix versions for frequently request capabilities

Where change is not recent

martinsumner committed:
Default correction

martinsumner committed:
Merge branch 'openriak-3.4' into nhse-o34-orkv.i57-fixcapability

ThomasArts commented on src/riak_kv_put_fsm.erl:
```diff
@@ -189,17 +190,16 @@ get_put_coordinator_failure_timeout() ->
         app_helper:get_env(riak_kv, retry_put_coordinator_failure, true) andalso
         get_option(retry_put_coordinator_failure, Options, true),
-    case {AckOption, AckCap, RetryCoord} of
+    case {AckOption, ?CAP_PUTFSM_ACK, RetryCoord} of
```
When this now always is enabled, this second argument of the tuple can be left out completely. This simplifies the code quite a bit:

```
case RetryCoord andalso not is_pid(AckOption) of
  true -> {true, [{ack_execute, self()}| Options]};
  false > {false, Options}
```

martinsumner replied:
I've retained the history of the capability, but also simplified the case statement

ThomasArts commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -560,9 +561,11 @@ check_root_version(Root, Index, Version) when is_integer(Version) ->
-    Version.
+    end
+%     ;
+% check_root_version(_Root, _Index, Version) ->
```
Weird, you should get a type error here. Since `legacy` is a valid type for `version()` as output, it should also be valid as input and then the first clause won't match. So, that would be detectable with a type checker, one had hoped...

martinsumner replied:
Clarified types

ThomasArts reviewed (approved):
It might be a good thing to keep the old defaults as comment in the `riak_kv_capability.hrl` header file. On the other hand, newcomers might be confused by seeing that some capabilities have been configurable before but are fixed now.

Should you also remove schema definitions for those capabilities that you cannot set any more?

martinsumner committed:
Simplify code and clarify types to represent fixed capability

Retains the history of the capability, but makes it clearer in types and code about the change in behaviour

martinsumner commented:
> It might be a good thing to keep the old defaults as comment in the `riak_kv_capability.hrl` header file. On the other hand, newcomers might be confused by seeing that some capabilities have been configurable before but are fixed now.
> 
> Should you also remove schema definitions for those capabilities that you cannot set any more?

Object format is the only configurable one, and that remains configurable
